libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -529,6 +538,11 @@ method rpcHandler*(g: GossipSub,
 
       libp2p_gossipsub_duplicate.inc()
 
+      #Dont relay to the peers from which we already received 
+      #We do it for large messages only
+      if msg.data.len > msgId.len * 10:     
```

Comment:
if the message is in `seen`, it should not be sent to anyone meaning we should not need to track it individually (doing so costs significant amounts of memory)

Reply:
actually, right - the seen check has been done and we need to stop the message after the seen check . hm . it would likely be better that we keep track of in-flight / queued messages and cancel these sends specifically on idontwant rather than storing all idontwant hashes

Reply:
Don't we already have this here https://github.com/vacp2p/nim-libp2p/blob/0911cb20f4a0b880451938988f24cb077cfd8b1b/libp2p/protocols/pubsub/gossipsub.nim#L319 ?

Reply:
> actually, right - the seen check has been done and we need to stop the message after the seen check . hm . it would likely be better that we keep track of in-flight / queued messages and cancel these sends specifically on idontwant rather than storing all idontwant hashes

Haven't this been already tried here? https://github.com/vacp2p/nim-libp2p/pull/851#issuecomment-1586818191

Reply:
I believe `heDontWants` is meant to handle `idontwant` msgs received and we already have the seen data structure to avoid relaying duplicates.

Reply:
We could check the `idowant` again when removing a msg from the non-prio queue. We would need to:

1.  Store the msg ID in the queue.
2. Have a `ref` to the structure in the `Pubsubpeer` object.

Reply:
> actually, right - the seen check has been done and we need to stop the message after the seen check . hm . it would likely be better that we keep track of in-flight / queued messages and cancel these sends specifically on idontwant rather than storing all idontwant hashes

Yes, for 12 (1MB) message transmissions in a 1000 peer network, we cancel roughly 8k-10k messages [here](https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub.nim#L412-L419). But rechecking for received idontwants before [actual transmission](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R348-R353) eliminates additional 1K-2K messages. [In-flight message elimination](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R355-R365) also filters 6K-8K messages, but benefits of in-flight message cancellation are questionable IMO

Reply:
> I believe `heDontWants` is meant to handle `idontwant` msgs received and we already have the `seen` data structure to avoid relaying duplicates.

IMO we can use heDontWants just before forwarding message to the peer to check if we have already received IDontWant [here](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R346-R353)

Reply:
> We could check the `idontwant` again when removing a msg from the non-prio queue. We would need to:
> 
> 1. Store the msg ID in the queue.

Yes, its done [here](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R70-R78)

> 2. Have a `ref` to the structure in the `Pubsubpeer` object.

I'm [passing message ID in this case](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R331)

Reply:
> > We could check the `idontwant` again when removing a msg from the non-prio queue. We would need to:
> > 
> > 1. Store the msg ID in the queue.
> 
> Yes, its done [here](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R70-R78)
> 
> > 2. Have a `ref` to the structure in the `Pubsubpeer` object.
> 
> I'm [passing message ID in this case](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R331)

Yes, I commented before reading the whole PR. I deleted the comment but you were still able to reply to it.

Reply:
> > > I believe `heDontWants` is meant to handle `idontwant` msgs received and we already have the `seen` data structure to avoid relaying duplicates.
> > 
> > 
> > IMO we can use heDontWants just before forwarding message to the peer to check if we have already received IDontWant [here](https://github.com/vacp2p/nim-libp2p/pull/1100/files#diff-cd23f9cd2a0f9c037ef0b4245ee7d06c09a24985c8ee5ed39fff496f705570c3R346-R353)
> 
> Yes, that's what I meant. But here it seems to me that you are adding a message id to the peer list even though the peer didn't send an `idontwant` for this id. That's not the purpose of the data structure.

If we received a message from a peer, it implies that peer does not want that message (we can kind of reuse hedontwants data structure for faster elimination of messages). otherwise we'll have to check both, seen cache and hedontwants

Reply:
I believe we already don't send a message from a peer we've just received it and also don't relay the same msg twice. Isn't this equivalent?

Reply:
This data structure is probably to store ID of msgs other peers in the network have already seen but we haven't yet. Then when we do see it, we don't send it to them.

Reply:
IMO our send inserts the message in non-priority queue, and may consume some time before its aired. if we receive the message during this interval, we can still cancel transmit

Reply:
True, but https://github.com/vacp2p/nim-libp2p/blob/0911cb20f4a0b880451938988f24cb077cfd8b1b/libp2p/protocols/pubsub/gossipsub.nim#L319 and checking `idontwant` when dequeuing from non-prio are enough. We don't seem to need this code for that.

Reply:
Maybe we could even end up not sending the msg to the current peer as it is enqueued and before sending we check `idontwant`.

Reply:
yes, this part does not have much impact!